### ISSUE TITLE
Windows: WorkArea property not updated

### ISSUE THREAD

**Issue by Pettor** [electron/electron#6312]
- Electron version: 1.x and above
- Operating system: Windows 10

I was previously using version 0.37.3 for my application and tried to upgrade to 1.0 but this caused the WorkArea property of [screen](https://github.com/electron/electron/blob/master/docs/api/screen.md) to not get updated to desktop WorkArea changes while the application is running.

For example. My resolution is 2560 \* 1440 so if the taskbar is located at the bottom of the desktop the WorkArea property will look like `{ x: 0, y: 40, width: 2560, height: 1400 }` (indicating the taskbar is 40pixels heigh and located at the bottom). If I locate the taskbar area to left instead the property will be `{ x: 82, y: 0, width: 2478, height: 1440 }` indicating the taskbar is located to the left. In version 0.37.3 I could use this logic to perfectly place the application over the tray icon depending on the taskbar location.

With newer Electron version the WorkArea property never changes. It's always returns the values according to what the WorkArea are when the application started. So if I start my application with the taskbar located at the bottom the WorkArea property will always be `{ x: 0, y: 40, width: 2560, height: 1400 }` even if the taskbar is changed to be located to the left. If I restart the application the property gets updated to show that the taskbar is now to the left (`{ x: 82, y: 0, width: 2478, height: 1440 }`).

The `display-metrics-changed` event of screen doesn't respond to taskbar changes either.

**Comment by miniak** [comment]
This bug is also present on macOS. The `display-metrics-changed` event does not fire when the Dock position changes, which should affect the work area.

**Comment by z-ax** [comment]
Yeah, and to this issue as well goes next one, but could split up if you want.
When another monitor was plugged, the `display.workArea` (from `screen.getPrimaryDisplay()`) is returned corrupted - `display.workArea` has actually the same bounds as `display.size`, so can't determine where the taskbar is (on Windows).

We have a tray application, so this issue is a major one to actually position our window correctly.

**Comment by z-ax** [comment]
Curious thing is when 2 monitors are plugged before app's launch - all area bounds come as expected.

**Comment by ykulinichenko** [comment]
The same bug is reproducing for me.

**Comment by KulykOleg** [comment]
Chacked 1.2.6 version. Same bug reproducing .

**Comment by z-ax** [comment]
Checked with v1.3.0 - still an issue.

**Comment by kevinsawicki** [comment]
> The display-metrics-changed event of screen doesn't respond to taskbar changes either.

I'm seeing the same behavior in standalone Chrome 52.0.2743.82 when calling [`chrome.system.display.getInfo`](https://developer.chrome.com/apps/system_display#method-getInfo) from a devtools extension. It never updates the `x` and `y` values after the taskbar changes position.

@Pettor Did this work for you in Electron 0.37.3?

**Comment by wenkaili** [comment]
I'm seeing a related issue with the WorkArea property. When I start the application under a certain resolution (1920x1080 for example), the WorkArea property is getting the height and width correctly. However, when I switch to a different resolution and switch back (all while the app is still running), the WorkArea property now returns a slightly different height value. 

I'm on Electron 1.2.7.

**Comment by zcbenz** [comment]
By reading Chromium's source code, I found this is caused by Chromium caching the results of displays, while not updating them when taskbar changes. So this is essentially a bug of Chromium, and fixing it also requires patching Chromium.

**Comment by z-ax** [comment]
@zcbenz Thanks for checking it out!
So, there is no way right now to handle events for that to at least relaunch the app if positioning on screen is essential (for apps like tray popup)?

**Comment by Pettor** [comment]
@kevinsawicki Sorry for late reply but currently using 0.37.3 and it behaves correctly. Only property that is static until logout / login is the scale factor, but that could be a Windows issue.

**Comment by z-ax** [comment]
Is there an issue in Chromium for this regression? Didn't find.

**Comment by Ciantic** [comment]
For windows, it might be possible to use FFI and get a display workArea independently of Chromium API. 

I'm looking in to this, the workArea goes haywire as I resume the computer from sleep, and it stays funny until I restart my app.

**Comment by llhupp** [comment]
@z-ax I looked around but was unable to find a related Chromium issue. I'm surprised this has been an issue for 5 months without something ticketed on their side... Is there someone with more knowledge of Electron's interface with Chromium that could file an issue related to this and get more visibility on this?

**Comment by Ciantic** [comment]
```
let getWorkArea = (point: {x: number, y: number}) => {
    let Point = struct({
        'x' : 'long',
        'y' : 'long',
    });
    let pointPtr = new Point(point);
    let hmonitor = User32.MonitorFromPoint(pointPtr.ref(), 2); // MONITOR_DEFAULTTONEAREST == 2
    let dpiXptr = ref.alloc('int');
    let dpiYptr = ref.alloc('int');
    Shcore.GetDpiForMonitor(hmonitor, 0, dpiXptr, dpiYptr);
    let dpiX =  dpiXptr.readInt32LE(0);
    let dpiY = dpiYptr.readInt32LE(0);
    let Rect = struct({
        'left' : 'long',
        'top' : 'long',
        'right' : 'long',
        'bottom' : 'long',
    });
    let rectPtr = new Rect();
    User32.SystemParametersInfoW(48, 0, rectPtr.ref(), 0); // SPI_GETWORKAREA 0x0030 == 48
    return {
        x : 0,
        y : 0,
        width : Math.ceil(rectPtr.right * 96 / dpiX),
        height : Math.ceil(rectPtr.bottom * 96 / dpiY)
    };
}
app.on('ready', () => {
    let display = screen.getDisplayNearestPoint({
        x : 0,
        y : 0
    });
    let workArea = getWorkArea({
        x : 0,
        y : 0
    });
    console.log("display?", display.workArea);
    console.log("workarea winapi?", workArea);
    app.quit();
});

**Comment by Pettor** [comment]
@llhupp Agree, haven't found any Chromium issue and sadly don't know the details of Electron APIs good enough to write a proper issue.

@zcbenz Can you write a Chromium issue or link to existing one for this?

@Ciantic Thanks! This is what I was thinking of doing as well. Given up hope that this will be solved any time soon and can't upgrade Electron until this has been fixed.

**Comment by zanderz** [comment]
It looks like Chromium does not track changes to screen work areas properly, and returns the stale values still to Electron. I opened a Chromium bug and am testing a change: https://bugs.chromium.org/p/chromium/issues/detail?id=712871

I could only get it working with electron/libchromiumcontent, which seems to lag Chrome's master by a  couple of months.

**Comment by zanderz** [comment]
The PR was merged into Chromium last night: https://codereview.chromium.org/2829693003/
note this is a Windows-only fix.

**Comment by kevinsawicki** [comment]
Thanks so much for fixing this @zanderz, your fix will be included in the next Electron release, 1.7.2

**Comment by Pettor** [comment]
I have tried 1.7.5 beta and it works as expected. Thx and gj 👍
